File: pa5/List.c
// List.c
// Eric Yup
// eyup@ucsc.edu
// pa5
// List ADT implementation

#include<stdio.h>
#include<stdlib.h>
#include "List.h"

// structs --------------------------------------------------------------------

// private NodeObj type
typedef struct NodeObj{
 int data;
 struct NodeObj* next;
 struct NodeObj* prev;
} NodeObj;

// private Node type
typedef NodeObj* Node;

// private ListObj type
typedef struct ListObj{
 Node head;
 Node tail;
 Node cursor;
 int length;
} ListObj;
typedef ListObj* List;

// Constructors-Destructors ---------------------------------------------------

Node newNode(int data){
 Node N = malloc(sizeof(NodeObj));
 N->data = data;
 N->next = NULL;
 N->prev = NULL;
 return(N);
}

void freeNode(Node* pN){
 if( pN!=NULL && *pN!=NULL ){
  free(*pN);
  *pN = NULL;
}
}

List newList(void){
 List L;
 L = malloc(sizeof(ListObj));
 L->head = L->tail = L->cursor = NULL;
 L->length = 0;
 return(L);
}

void freeList(List* pL){
 if( pL!=NULL && *pL!=NULL ){
  if(length(*pL) > 0 )
   clear(*pL);
 free(*pL);
 *pL = NULL;
}
}

// Access functions -----------------------------------------------------------
int length(List L){
 if( L == NULL ){
  printf("List Error: calling length() on NULL List reference\n");
  exit(1);
}
int length = 0;
if (L->head == NULL){
  return length;
}
Node M = L->head;
while(M != NULL){
  length++;
  M = M->next;
}
return length;
}

int index(List L){
  if(L->cursor == NULL) return -1;
  Node temp = L->head;
  int ind = 0;
  while( temp != NULL ){
   if( temp->data == L->cursor->data) return ind;
   ind++;
   temp= temp->next;
 }
 return -1;

}


int front(List L){
  if( L==NULL ){
    printf("List Error: calling front() on NULL List reference\n");
    exit(1);
  }
  if( length(L) > 0) 
   return L->head->data;
 return -1;
}

int back(List L){
	if( L==NULL ){
    printf("List Error: calling back() on NULL List reference\n");
    exit(1);
  }
  if ( length(L) > 0) 
   return L->tail->data;
 return -1;
}

int get(List L){
	if(L == NULL){
    fprintf(stderr, "List Error: calling get() on NULL List reference\n");
      exit(0); //exit out of this stamtement if true  
    }
    if(length(L) > 0 && index(L) >= 0) return L->cursor->data;
    return -1;
  }

  int equals(List A, List B){
    if( length(A) != length(B)) return 0;
    Node tempA = A->head;
    Node tempB = B->head;

    while( tempA != NULL){
     if( tempA->data != tempB->data) return 0;
     tempA = tempA->next;
     tempB = tempB->next;
   }
   return 1;
 }

// Manipulation procedures ----------------------------------------------------
 void clear(List L){
  L->cursor = NULL;
  while(length(L)>0){
    deleteFront(L);
  }
  L->length = 0;
}

void moveFront(List L){
	if(length > 0)
		L->cursor = L->head;
}

void moveBack(List L){
	if(length > 0)
   L->cursor = L->tail;
}

void movePrev(List L){
  if(index(L) == -1) 
   return;
 if(index(L) == 0) 
  L->cursor = NULL;
if ( index(L) > 0 && index(L) <= length(L) -1)
 L->cursor = L->cursor->prev;
}

void moveNext(List L){
  if(index(L) == -1) 
   return;
 if(index(L) == length(L)-1) 
   L->cursor = NULL;
 if (index(L) >= 0 && index(L) <= length(L) -1) 
   L->cursor = L->cursor->next;
}

void prepend(List L, int data){
  Node temp = newNode(data);
  if(length(L) == 0){
   L->head = L->tail = temp;   
   return;
 }           
 temp->next = L->head;
 L->head->prev = temp;
 L->head = L->head->prev;
}

void append(List L, int data){
  Node temp = newNode(data);
  if(length(L) == 0){
   L->head = L->tail = temp;   
   return;
 }  
 temp->prev = L->tail;
 L->tail->next = temp;
 L->tail = L->tail->next;
}

void insertBefore(List L, int data){  
  if(length(L)>0 && index(L)>=0){
   if(index(L)== 0){
    prepend(L,data);
    return;
  }
  Node ins = newNode(data);
  ins->next = L->cursor;
  ins->prev = L->cursor->prev;
  L->cursor->prev->next = ins;
  L->cursor->prev = ins;
}
}

void insertAfter(List L, int data){
  if(length(L)>0 && index(L)>=0){ 
   if(index(L)== length(L)-1){
    append(L,data);
    return;
  }  
  Node ins = newNode(data);                      
  ins->next = L->cursor->next;  
  ins->prev = L->cursor;        
  L->cursor->next->prev = ins;
  L->cursor->next = ins;
}
}

void deleteFront(List L){
  if( length(L) < 0) return;
  if(index(L) == 0)
   L->cursor = NULL;
 if(length(L) == 1){
  Node delF = L->head;
  L->head = NULL;
  L->tail= NULL;
  freeNode(&delF);
  return;
}
Node delF = L->head;
L->head = L->head->next;
L->head->prev = NULL;
freeNode(&delF);
}

void deleteBack(List L){
  if( length(L) < 0) return;
  if(index(L) == length(L)-1)
   L->cursor = NULL;
 if( length(L) == 1){
         //clear(L);
   Node delB = L->tail;
   L->head = NULL;
   L->tail= NULL;
   freeNode(&delB);
         //delB = NULL;
   return;
 }
 Node delB = L->tail;
 L->tail = L->tail->prev;
 L->tail->next = NULL;
 freeNode(&delB);
}

void delete(List L){
 Node del = L->cursor;
 if(length(L)== 0){
  printf("List Error: calling delete() on empty list\n");
  exit(1);
}
if(L->head == L->cursor){
  deleteFront(L);
  return;
}
if(L->tail == L->cursor){
  deleteBack(L);
  return;
}
if(length(L)>0 && index(L)>=0){      
 if(del != NULL){
  L->cursor->prev->next = L->cursor->next;
  L->cursor->next->prev = L->cursor->prev;
  L->cursor = NULL;
}
}
freeNode(&del);
}

// Other operations -----------------------------------------------------------
void printList(FILE* out, List L){
  Node traverse = L->head;
  while( traverse != NULL){
    fprintf(out,"%d ", traverse->data);
    traverse = traverse->next;
  }
}

List copyList(List L){
  List N = newList();
  Node traverse = L->head;

  while( traverse!=NULL ){
   append(N, traverse->data);
   traverse = traverse->next;
 }
 return N;
}

File: pa3/List.java
// List.java
// Eric Yup
// eyup@ucsc.edu
// pa3 
// doubly linked list ADT
public class List {
	// Node subclass
	private class Node {
		Object data;
		Node next;
		Node prev;

		// Constructor for Node
		Node(Object data) {
			this.data = data;
			next = null;
			prev = null;
		}

		// toString: Overides Object's toString method.
		public String toString() {
			return String.valueOf(data);
		}
	}

	private Node head;
	private Node tail;
	private int length;
	private Node cursor;

	// Constructor
	// Creates new list.
	List() {
		head = null;
		tail = null;
		length = 0; // initialize
		cursor = null;
	}

	// Access functions
	int length() {
		return length;
	}// Returns the number of elements in this List.

	int index() {
		if (cursor == null)
			return -1;
		Node temp = head;
		int index = 0;
		while (temp != null) {
			if (temp.data == cursor.data)
				return index;
			index++;
			temp = temp.next;
		}
		return -1;
	} // If cursor is defined, returns the index of the cursor element,
		// otherwise returns -1.

	Object front() {
		if (length() > 0)
			return head.data;
		return -1;
	}

	// Returns front element. Pre: length()>0
	Object back() {
		if (length() > 0)
			return tail.data;
		return -1;// Returns back element. Pre: length()>0
	}

	Object get() {
		if (length() > 0 && index() >= 0)
			return cursor.data;
		return -1;
	}
	// Returns cursor element. Pre: length()>0, index()>=0

	public boolean equals(Object x) {
		boolean flag = true;
		if (x instanceof List) {
			List L = (List) x;
			Node A = this.head;
			Node B = L.head;
			if (this.length == L.length) {
				while (flag && A != null) {
					flag = (A.data.equals(B.data));
					A = A.next;
					B = B.next;
				}
				return flag;
			} else {
				return false;
			}
		} else {
			return false;
		}
	}

	// Returns true if and only if this List and L are the same
	// integer sequence. The states of the cursors in the two Lists
	// are not used in determining equality.
	// Manipulation procedures
	void clear() {
		head = null;
		tail = null;
		length = 0;
		cursor = null;
	}

	// Resets this List to its original empty state.
	void moveFront() {
		if (length() > 0) {
			cursor = head;
		}
	}

	// If List is non-empty, places the cursor under the front element,
	// otherwise does nothing.
	void moveBack() {
		if (length() > 0) {
			cursor = tail;
		}
	}// .next??
	// If List is non-empty, places the cursor under the back element,
	// otherwise does nothing.

	void movePrev() {
		if (index() == -1) {
			return;
		}
		if (index() == 0) {
			cursor = null;
		}
		if (index() >= 0 && index() <= length() - 1) {
			cursor = cursor.prev;
		} // If cursor is defined and not at front, moves cursor one step toward
	}
	// front of this List, if cursor is defined and at front, cursor becomes

	// undefined, if cursor is undefined does nothing.

	void moveNext() {
		if (index() == -1) {
			return;
		}
		if (index() == length() - 1) {
			cursor = null;
		}
		if (index() >= 0 && index() <= length() - 1) {
			cursor = cursor.next;
		}
	}

	// If cursor is defined and not at back, moves cursor one step toward
	// back of this List, if cursor is defined and at back, cursor becomes
	// undefined, if cursor is undefined does nothing.
	void prepend(Object data) {
		List.Node temp = new List.Node(data);
		if (length() > 0) {
			temp.next = head;
			head.prev = temp;
			head = head.prev;
		}
		if (length() == 0) {
			head = temp;
			tail = temp;
		}
		length++;
	}

	// Insert new element into this List. If List is non-empty,
	// insertion takes place before front element.
	void append(Object data) {
		List.Node temp = new List.Node(data);
		if (length() > 0) {
			tail.next = temp;
			temp.prev = tail;
			tail = tail.next;
		}
		if (length() == 0) {
			head = temp;
			tail = temp;
		}
		length++;
	}

	// Insert new element into this List. If List is non-empty,
	// insertion takes place after back element.
	void insertBefore(Object data) {
		if (length() > 0 && index() >= 0) {
			if (index() == 0) {
				prepend(data);
				return;
			}
			List.Node temp = new List.Node(data);
			temp.next = cursor;
			temp.prev = cursor.prev;
			cursor.prev.next = temp;
			cursor.prev = temp;
		}
		length++;
	}

	// Insert new element before cursor.
	// Pre: length()>0, index()>=0
	void insertAfter(Object data) {
		if (length() > 0 && index() >= 0) {
			if (index() == length() - 1) {
				append(data);
				return;
			}
			List.Node temp = new List.Node(data);
			temp.next = cursor.next;
			temp.prev = cursor;
			cursor.next.prev = temp;
			cursor.next = temp;
		}
		length++;
	}

	// Inserts new element after cursor.
	// Pre: length()>0, index()>=0
	void deleteFront() {
		if (length() < 0) {
			return;
		}
		if (index() == 0)
			cursor = null;
		if (length() == 1) {
			clear();
			return;
		}
		head = head.next;
		head.prev = null;
		length--;
	}

	// Deletes the front element. Pre: length()>0
	void deleteBack() {
		if (length() < 0)
			return;
		if (index() == length() - 1)
			cursor = null;
		if (length() == 1) {
			clear();
			return;
		}
		tail = tail.prev;
		tail.next = null;
		length--;
	}

	// Deletes the back element. Pre: length()>0
	void delete() {
		List.Node temp = cursor;
		if (length() == 0) {
			return;
		}
		if (head == cursor) {
			deleteFront();
			return;
		}
		if (tail == cursor) {
			deleteBack();
			return;
		}
		if (length() > 0 && index() >= 0) {
			if (temp != null) {
				cursor.prev.next = cursor.next;
				cursor.next.prev = cursor.prev;
				cursor = null;
			}
		}
		length--;
	}

	// Deletes cursor element, making cursor undefined.
	// Pre: length()>0, index()>=0
	// Other methods
	public String toString() {
		String str = "";
		List.Node temp = head;
		while (temp != null) {
			if (temp == tail) {
				str += temp.data;
				temp = temp.next;
				break;
			}
			str += temp.data + " ";
			temp = temp.next;
		}
		return str;
	}

	// Overrides Object's toString method. Returns a String
	// representation of this List consisting of a space
	// separated sequence of integers, with front on left.
	public List copy() {
		List C = new List();
		List.Node traverse = head;
		while (traverse != null) {
			C.append(traverse.data);
			traverse = traverse.next;
		}
		return C;
	}
	// Returns a new List representing the same integer sequence as this
	// List. The cursor in the new list is undefined, regardless of the
	// state of the cursor in this List. This List is unchanged.

}